CanvasPlusPlayground/Features/ToDoItems/ToDoListManager.swift

```
@@ -0,0 +1,90 @@
+                    onCacheReceive: { cached in
+                        guard let cached = cached?.first else { return }
+                        self.toDoItemCount = cached.assignmentsNeedingSubmitting
+                    }
+                )
+
+            // TODO: If we support grading assignments, add that count
+            self.toDoItemCount = count?.first?.assignmentsNeedingSubmitting
+        } catch {
+            LoggerService.main.error("Failed to fetch to-do item count: \(error)")
+        }
+    }
+
+    func fetchToDoItems(courses: [Course]) async {
+        let request = CanvasRequest.getToDoItems(include: [.ungradedQuizzes])
+
+        var newItems = [ToDoItem]()
+
+        do {
+            let _: [ToDoItem]? = try await CanvasService.shared
+                .loadAndSync(
+                    request,
+                    onCacheReceive: { cached in
+                        guard let cached else { return }
+                        toDoItems = []
+                        self.addItems(cached, to: &toDoItems, courses: courses)
+                    },
+                    loadingMethod: .all(onNewPage: { items in
+                        self.addItems(items, to: &newItems, courses: courses)
+                    })
+                )
+        } catch {
+            LoggerService.main.error("Failed to fetch to-do items: \(error)")
+        }
+
+        toDoItems = newItems
+    }
+
+    func ignoreToDoItem(_ item: ToDoItem) async {
+        // We can either use the ignoreURL or the ignorePermanentlyURL.
+        // ignoreURL will add the item back if the item is updated in the future.
+        // ignorePermanentlyURL will remove the item from the list forever.
+
+        let request = CanvasRequest.ignoreToDoItem(ignoreURL: item.ignoreURL)
+
+        do {
+            try await CanvasService.shared.fetch(request)
+        } catch {
+            LoggerService.main.error("Failed to ignore todo item: \(error)")
+        }
+    }
```

can we propagate the error to the ui somehow, u can decide how

---

can we move this to next build? I'll make an issue

---

sure